src/workerd/api/node/node.h

```diff
@@ -23,7 +24,51 @@ class CompatibilityFlags : public jsg::Object {
   CompatibilityFlags() = default;
   CompatibilityFlags(jsg::Lock&, const jsg::Url&) {}
 
+  // The compatibility flags method here is called at most once per isolate realm.
+  // After that, the flags are cached and always readonly.
+  jsg::JsObject getCompatibilityFlags(jsg::Lock& js, workerd::CompatibilityFlags::Reader flags) {
+    auto obj = js.objNoProto();
+    auto dynamic  = capnp::toDynamic(flags);
+    auto schema = dynamic.getSchema();
+
+    auto on = js.strIntern("on");
+    auto off = js.strIntern("off");
+    bool skipExperimental = !flags.getWorkerdExperimental();
+
+    for (auto field : schema.getFields()) {
+      // If this is an experimental flag, we expose it only if the experimental mode
+      // is enabled.
+      auto annotations = field.getProto().getAnnotations();
+      bool skip = false;
+      if (skipExperimental) {
+        for (auto annotation : annotations) {
+          if (annotation.getId() == EXPERIMENTAl_ANNOTATION_ID) {
+            skip = true;
+            break;
+          }
+        }
+      }
+      if (skip) continue;
+
+      bool value = dynamic.get(field).as<bool>();
+      for (auto annotation : annotations) {
+        if (annotation.getId() == COMPAT_ENABLE_FLAG_ANNOTATION_ID) {
+          obj.setReadOnly(js, annotation.getValue().getText(), value ? on : off);
+        }
+        else if (annotation.getId() == COMPAT_DISABLE_FLAG_ANNOTATION_ID) {
+          // Take note of the inverted ternary here... above it is on : off, here it is off : on
+          obj.setReadOnly(js, annotation.getValue().getText(), value ? off : on);
+        }
+      }
+    }
+
+    obj.seal(js);
+    return obj;
+  }
+
   JSG_RESOURCE_TYPE(CompatibilityFlags, workerd::CompatibilityFlags::Reader flags) {
+    JSG_LAZY_READONLY_INSTANCE_PROPERTY(compatibilityFlags, getCompatibilityFlags);
```

Comment:
Does this mean the result will always be the same, hence caching?

Reply:
essentially yes. If this were `JSG_READONLY_INSTANCE_PROPERTY`, then the getter would be called every time the property is accessed, which is wasteful. By using the lazy option, v8 will call the getter only once and then will replace the getter in the internal property definition with the returned value so the getter is never called again.

Reply:
Ah nice. Thanks for the explanation!